### PR TITLE
Feature/prod deploy

### PR DESCRIPTION
**Pull request by anthonymartin**
* update comments 
* adds extra events to emit
* changed hardcoded 4 to be dynamic
* removes deployments directory
* adds core and frontend tags to relevant deploy scripts 
* adds butter address
* removes warnings from solidity compilation for ButterBatchProcessing
* adds staking contracts to contract registry in deploy scripts
* add spdx identifier and docgen solc hinter to all contract files
* updates RewardsEscrow deployment to use mainnet pop address

### CODE REVIEW

File: packages/hardhat/contracts/core/utils/ACLRegistry.sol [hunk 1]
```diff
@@ -54,6 +55,9 @@ contract ACLRegistry is IACLRegistry {
   mapping(bytes32 => address) private _permissions;
 
   bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+  bytes32 public constant KEEPER_ROLE = keccak256("Keeper");
+  bytes32 public constant DAO_ROLE = keccak256("DAO");
+  bytes32 public constant APPROVED_CONTRACT_ROLE = keccak256("ApprovedContract");
```
Comment (RedVeil) [inline packages/hardhat/contracts/core/utils/ACLRegistry.sol:60]: I really like this change

File: packages/hardhat/lib/tasks/acl/grantRole.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { ethers } from "ethers";
+import { task } from "hardhat/config";
+interface Args {
+  role: string;
+  address: string;
+}
+export default task("acl:grant-role", "check if address has role")
```
Comment (ecmendenhall) [inline packages/hardhat/lib/tasks/acl/grantRole.ts]:
```suggestion
export default task("acl:grant-role", "grant ACLRegistry role to address")
```

File: packages/hardhat/lib/tasks/acl/grantRole.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  .addParam("address", "account to check")
```
Comment (ecmendenhall) [inline packages/hardhat/lib/tasks/acl/grantRole.ts]:
```suggestion
  .addParam("address", "address to be granted role")
```

File: packages/hardhat/lib/tasks/acl/grantRole.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  .addParam("role", "role to check")
```
Comment (ecmendenhall) [inline packages/hardhat/lib/tasks/acl/grantRole.ts]:
```suggestion
  .addParam("role", "name of role to grant")
```

File: packages/hardhat/lib/tasks/keeper-incentive/update.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import { parseEther } from "ethers/lib/utils";
+import { task } from "hardhat/config";
+interface Args {
+  contract: string;
+  index: string;
+  reward: string;
+  enabled: string;
+  permissionless: string;
+}
+export default task("keeper-incentive:update", "creates keeper incentive")
```
Comment (ecmendenhall) [inline packages/hardhat/lib/tasks/keeper-incentive/update.ts]:
```suggestion
export default task("keeper-incentive:update", "updates keeper incentive")
```

File: packages/hardhat/lib/tasks/keeper-incentive/update.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  .addParam("contract", "name of contract")
+  .addParam("index", "index of incentive")
+  .addParam("reward", "reward amount")
```
Comment (ecmendenhall) [inline packages/hardhat/lib/tasks/keeper-incentive/update.ts]: in wei, I assume?

```suggestion
  .addParam("reward", "reward amount (wei)")
```

File: packages/hardhat/lib/utils/constants.ts [hunk 7]
```diff
@@ -6,3 +6,5 @@ export const networkMap = {
   137: "polygon",
   1337: "hardhat",
 };
+
+export const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
```
Comment (ecmendenhall) [inline packages/hardhat/lib/utils/constants.ts]: Can we use `ethers.constants.AddressZero`?

File: packages/hardhat/lib/tasks/keeper-incentive/update.ts (hunk 6 above)
Comment (anthonymartin) [reply packages/hardhat/lib/tasks/keeper-incentive/update.ts:14]: the reward amount is converted to ether below

File: packages/hardhat/lib/utils/constants.ts (hunk 7 above)
Comment (anthonymartin) [reply packages/hardhat/lib/utils/constants.ts:10]: ah yes, I had been looking for that const when I added this one, but I couldn't find it

File: packages/hardhat/deploy/utils.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    ethers.utils.id(contractName)
+  );
+
+  if (contract === ZERO_ADDRESS) {
+    await contractRegistry.addContract(
+      ethers.utils.id(contractName),
+      (
+        await deployments.get(contractName)
+      ).address,
+      ethers.utils.id("1"),
```
[the remark this answers is not in the document]
Comment (anthonymartin) [reply packages/hardhat/deploy/utils.ts:28]: this only runs if it's the first version, so I don't think that's necessary

**Pull request merged**